From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Natalie Vock <natalie.vock@gmx.de>
Date: Sat, 20 Jun 2026 13:09:19 +0200
Subject: [PATCH 12/17] STEAMOS: radv: VK_EXT_pageable_device_local_memory

This relies on a feature that is only present in SteamOS kernels. When
running under an incompatible kernel, the ioctl will fail and priority
manipulation becomes a no-op.

(cherry picked from commit ced99f20ee2cf49c0c8bfc3a8483f7a2f97985d2)
(cherry picked from commit 2bda033d6a8557d104355ca5248edf52391efc48)
---
 include/drm-uapi/amdgpu_drm.h                 |  1 +
 src/amd/vulkan/radv_device_memory.c           |  9 ++++++
 src/amd/vulkan/radv_physical_device.c         |  4 +++
 src/amd/vulkan/radv_radeon_winsys.h           |  2 ++
 src/amd/vulkan/winsys/amdgpu/radv_amdgpu_bo.c | 30 +++++++++++++++++++
 5 files changed, 46 insertions(+)

diff --git a/include/drm-uapi/amdgpu_drm.h b/include/drm-uapi/amdgpu_drm.h
index 1f70a3ff056..2afe62f35ef 100644
--- a/include/drm-uapi/amdgpu_drm.h
+++ b/include/drm-uapi/amdgpu_drm.h
@@ -806,6 +806,7 @@ union drm_amdgpu_wait_fences {
 #define AMDGPU_GEM_OP_GET_GEM_CREATE_INFO	0
 #define AMDGPU_GEM_OP_SET_PLACEMENT		1
 #define AMDGPU_GEM_OP_GET_MAPPING_INFO		2
+#define AMDGPU_GEM_OP_SET_PRIORITY		3
 
 struct drm_amdgpu_gem_vm_entry {
 	/* Start of mapping (in bytes) */
diff --git a/src/amd/vulkan/radv_device_memory.c b/src/amd/vulkan/radv_device_memory.c
index dc137cfa890..464d69fdbc8 100644
--- a/src/amd/vulkan/radv_device_memory.c
+++ b/src/amd/vulkan/radv_device_memory.c
@@ -412,3 +412,12 @@ radv_GetDeviceMemoryCommitment(VkDevice device, VkDeviceMemory memory, VkDeviceS
 {
    *pCommittedMemoryInBytes = 0;
 }
+
+VKAPI_ATTR void VKAPI_CALL
+radv_SetDeviceMemoryPriorityEXT(VkDevice _device, VkDeviceMemory _memory, float priority)
+{
+   VK_FROM_HANDLE(radv_device, device, _device);
+   VK_FROM_HANDLE(radv_device_memory, memory, _memory);
+
+   device->ws->buffer_set_priority(device->ws, memory->bo, priority);
+}
diff --git a/src/amd/vulkan/radv_physical_device.c b/src/amd/vulkan/radv_physical_device.c
index e23d3755745..1fdcb31c8ce 100644
--- a/src/amd/vulkan/radv_physical_device.c
+++ b/src/amd/vulkan/radv_physical_device.c
@@ -903,6 +903,7 @@ radv_physical_device_get_supported_extensions(const struct radv_physical_device
       .EXT_mutable_descriptor_type = true, /* Trivial promotion from VALVE. */
       .EXT_nested_command_buffer = true,
       .EXT_non_seamless_cube_map = true,
+      .EXT_pageable_device_local_memory = true,
       .EXT_pci_bus_info = true,
 #ifndef _WIN32
       .EXT_physical_device_drm = true,
@@ -1468,6 +1469,9 @@ radv_physical_device_get_features(const struct radv_physical_device *pdev, struc
       /* VK_EXT_pipeline_robustness */
       .pipelineRobustness = true,
 
+      /* VK_EXT_pageable_device_local_memory */
+      .pageableDeviceLocalMemory = true,
+
       /* VK_KHR_maintenance5 */
       .maintenance5 = true,
 
diff --git a/src/amd/vulkan/radv_radeon_winsys.h b/src/amd/vulkan/radv_radeon_winsys.h
index 6bdda446c68..5ad99d6d433 100644
--- a/src/amd/vulkan/radv_radeon_winsys.h
+++ b/src/amd/vulkan/radv_radeon_winsys.h
@@ -241,6 +241,8 @@ struct radeon_winsys {
    void (*buffer_destroy)(struct radeon_winsys *ws, struct radeon_winsys_bo *bo);
    void *(*buffer_map)(struct radeon_winsys *ws, struct radeon_winsys_bo *bo, bool use_fixed_addr, void *fixed_addr);
 
+   void (*buffer_set_priority)(struct radeon_winsys *ws, struct radeon_winsys_bo *bo, float priority);
+
    VkResult (*buffer_from_ptr)(struct radeon_winsys *ws, void *pointer, uint64_t size, unsigned priority,
                                struct radeon_winsys_bo **out_bo);
 
diff --git a/src/amd/vulkan/winsys/amdgpu/radv_amdgpu_bo.c b/src/amd/vulkan/winsys/amdgpu/radv_amdgpu_bo.c
index 3a61e04bffd..d288df65268 100644
--- a/src/amd/vulkan/winsys/amdgpu/radv_amdgpu_bo.c
+++ b/src/amd/vulkan/winsys/amdgpu/radv_amdgpu_bo.c
@@ -33,6 +33,32 @@
 
 static void radv_amdgpu_winsys_bo_destroy(struct radeon_winsys *_ws, struct radeon_winsys_bo *_bo);
 
+static uint32_t
+vk_prio_to_amdgpu(float vk)
+{
+   return (uint32_t)((double)vk * (double)(UINT32_MAX - 1u));
+}
+
+static void
+radv_amdgpu_winsys_bo_set_priority(struct radeon_winsys *_ws, struct radeon_winsys_bo *_bo, float prio)
+{
+   struct radv_amdgpu_winsys *ws = radv_amdgpu_winsys(_ws);
+   struct radv_amdgpu_winsys_bo *bo = radv_amdgpu_winsys_bo(_bo);
+   struct drm_amdgpu_gem_op args;
+
+   if (!bo->base.is_local) {
+      bo->priority = (int)(prio * 31.0f);
+      return;
+   }
+
+   memset(&args, 0, sizeof(args));
+   args.handle = bo->bo_handle;
+   args.op = AMDGPU_GEM_OP_SET_PRIORITY;
+   args.value = vk_prio_to_amdgpu(prio);
+
+   drmCommandWriteRead(ws->fd, DRM_AMDGPU_GEM_OP, &args, sizeof(args));
+}
+
 static int
 radv_amdgpu_bo_va_op(struct radv_amdgpu_winsys *ws, uint32_t bo_handle, uint64_t offset, uint64_t size, uint64_t addr,
                      uint32_t bo_flags, uint64_t internal_flags, uint32_t ops, uint32_t wait_count,
@@ -722,6 +748,9 @@ radv_amdgpu_winsys_bo_create(struct radeon_winsys *_ws, uint64_t size, unsigned
       }
    }
 
+   if (bo->base.is_local)
+      radv_amdgpu_winsys_bo_set_priority(&ws->base, &bo->base, priority / 31.0f);
+
    if (initial_domain & RADEON_DOMAIN_GTT)
       p_atomic_add(&ws->alloc_tracker->allocated_gtt, align64(bo->base.size, ws->info.gart_page_size));
 
@@ -1351,6 +1380,7 @@ radv_amdgpu_bo_init_functions(struct radv_amdgpu_winsys *ws)
    ws->base.buffer_create = radv_amdgpu_winsys_bo_create;
    ws->base.buffer_destroy = radv_amdgpu_winsys_bo_destroy;
    ws->base.buffer_map = radv_amdgpu_winsys_bo_map;
+   ws->base.buffer_set_priority = radv_amdgpu_winsys_bo_set_priority;
    ws->base.buffer_unmap = radv_amdgpu_winsys_bo_unmap;
    ws->base.buffer_from_ptr = radv_amdgpu_winsys_bo_from_ptr;
    ws->base.buffer_from_fd = radv_amdgpu_winsys_bo_from_fd;
